### PR TITLE
C++: Assorted minor doc improvements

### PR DESCRIPTION
**Pull request by geoffw0**
Assorted minor doc improvements.  Most of the changes are intended to provide better context for autofix (e.g. clearer examples), or just to correct mistakes noticed while working on triage.  These changes should benefit human readers as well as autofix.

I'm happy to explain the thinking behind any particular change where it is not clear.

### CODE REVIEW

**Review by MathiasVP — DISMISSED** [review]
Changes LGTM! Now we just need to make CI happy 😄

**Review by MathiasVP — APPROVED** [review]

File: cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp
Comment (geoffw0) [reply cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp:259]: As far as I can tell there is no guarantee that `std::memcpy` _doesn't_ throw in the language standard, but most implementations won't.  The one I'm looking appears to get marked `nothrow` if it's compiled as C++, but not if it's compiled as C, and it appears to get C linkage and not `nothrow` in the actual database.  It also has a body that just calls a builtin function.

So we could perhaps address this by assuming that a C linkage function doesn't throw ... but I don't think it's actually strictly true that a C linkage function can't throw.  Or that builtin function don't throw, but again I'm not convinced that's true.  Which leaves us with modelling ... we could model functions that don't throw, but again I don't know if it's strictly true that all versions of `memset` don't throw, and it seems a bit heavy handed to model it given that `nothrow` exists.

**Comment by geoffw0** [comment]
The integration test is still failing, and I don't think it's anything to do with the PR.  So I'll try merging in `main` again tomorrow and see if that fixes it.  We're also waiting for a docs review.

**Comment by MathiasVP** [comment]
> The integration test is still failing, and I don't think it's anything to do with the PR. So I'll try merging in `main` again tomorrow and see if that fixes it. We're also waiting for a docs review.

The integration tests are failing because you changed the alert message of `WrongTypeFormatArguments`. Try searching for `WrongTypeFormatArguments` in the CI log 😄

File: cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp:259]: `std::memcpy` is from the C standard library, so it cannot throw. But you're right that it's not marked as `noexcept` in the standard.

Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp:259]: And this is probably the relevant part of the standard:

Screenshot 2024-07-10 at 14 07 22

So since `memcpy` cannot throw in C, and the contents of `cstring` and `string.h` is identical, my reading is that `memcpy` also isn't allowed to throw 🤔

**Review by subatoi — APPROVED** [review]
Looks good—thank you!

**Comment by geoffw0** [comment]
> The integration tests are failing because you changed the alert message of WrongTypeFormatArguments. Try searching for WrongTypeFormatArguments in the CI log. 

Ah, I didn't notice that.  I've made a semmle-code PR referencing this one, that fixes up the integration test.

File: cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp
Comment (geoffw0) [reply cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp:259]: You make a good case for modelling.  I'll create an issue for this.

Comment (geoffw0) [reply cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp:259]: Issue created.  Thanks for helping me get to the bottom of this!

**Comment by geoffw0** [comment]
Integration tests pass on the internal PR.  I'll merge both together once both are approved.

**Review by MathiasVP — APPROVED** [review]

**Review by MathiasVP — APPROVED** [review]

**Comment by geoffw0** [comment]
Thanks for merging when all was green. :)

**Comment by MathiasVP** [comment]
No problem! 😄

**Pull request merged**
